lib/staggered_kd_reorder_xinv.cu

```diff
@@ -0,0 +1,124 @@
+      // 2. Storing fineXinv, the reordered fine KD inverse field
+      return coarseXinv.Bytes() + fineXinv.Bytes();
+    }
+
+    unsigned int minThreads() const { return fineXinv.VolumeCB(); }
+
+  public:
+    CalculateStaggeredGeometryReorder(GaugeField& fineXinv, const GaugeField& coarseXinv, const double scale) :
+      TunableKernel3D(fineXinv, QUDA_KDINVERSE_GEOMETRY, 2),
+      fineXinv(fineXinv),
+      coarseXinv(coarseXinv),
+      scale(scale)
+    {
+      checkPrecision(fineXinv, coarseXinv);
+      checkLocation(fineXinv, coarseXinv);
+      if (fineXinv.Geometry() != QUDA_KDINVERSE_GEOMETRY)
+        errorQuda("Unsupported geometry %d", fineXinv.Geometry());
+      if (fineXinv.Ndim() != 4)
+        errorQuda("Number of dimensions %d is not supported", fineXinv.Ndim());
+      if (coarseXinv.Geometry() != QUDA_SCALAR_GEOMETRY)
+        errorQuda("Unsupported geometry %d", coarseXinv.Geometry());
+
+      strcat(aux,",computeStaggeredGeometryReorder");
+      if (dagger_approximation) strcat(aux, ",dagger_approximation");
+
+      apply(device::get_default_stream());
+    }
+
+    void apply(const qudaStream_t &stream)
+    {
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+      if (fineXinv.Location() == QUDA_CPU_FIELD_LOCATION) {
+        constexpr QudaGaugeFieldOrder fineOrder = QUDA_QDP_GAUGE_ORDER;
+        constexpr QudaGaugeFieldOrder coarseOrder = QUDA_QDP_GAUGE_ORDER;
+        CalculateStaggeredGeometryReorderArg<Float,fineColor,fineOrder,coarseOrder,dagger_approximation> arg(fineXinv, coarseXinv, scale);
+        launch_host<ComputeStaggeredGeometryReorder>(tp, stream, arg);
+
+      } else if (fineXinv.Location() == QUDA_CUDA_FIELD_LOCATION) {
+        constexpr QudaGaugeFieldOrder fineOrder = QUDA_FLOAT2_GAUGE_ORDER;
+        constexpr QudaGaugeFieldOrder coarseOrder = QUDA_MILC_GAUGE_ORDER;
+        CalculateStaggeredGeometryReorderArg<Float,fineColor,fineOrder,coarseOrder,dagger_approximation> arg(fineXinv, coarseXinv, scale);
+        launch_device<ComputeStaggeredGeometryReorder>(tp, stream, arg);
+
+      }
+    }
+  };
+
+  template<typename Float, int fineColor>
+  struct calculateStaggeredGeometryReorder {
+    calculateStaggeredGeometryReorder(GaugeField &fineXinv, const GaugeField &coarseXinv, const bool dagger_approximation, const double) {
+      // template on dagger approximation
+      if (dagger_approximation)  {
+        // approximate the inverse with the dagger: the free field for staggered, 
+        // B^-1 = 1 / (4 * (d + mass^2)), where the 4 is due to the factor of 2 convention
+        // Point of future investigation: it seems like we don't need the scaling factor to get
+        // a consistent norm?
+        double scale = 1.; // / (4. * (fineXinv.Ndim() + mass * mass));
+
+        // reset scales as appropriate
+        if constexpr (sizeof(Float) < QUDA_SINGLE_PRECISION) {
+          double max_scale = coarseXinv.abs_max() * abs(scale) * 1.01;
```

Comment:
What's the motivation for the 1% margin here?

Reply:
I remembered seeing bonus paddings elsewhere, but I can't seem to find them anymore. I'll zap it!

Reply:
https://github.com/lattice/quda/pull/1221/commits/494c0be0466700edb2329aeac14ced9a5ea7e9dd